libr/bin/obj.c

```diff
@@ -8,17 +8,135 @@
 	if (binfile->rbin->verbose) \
 	eprintf
 
+static void mem_free(void *data) {
+	RBinMem *mem = (RBinMem *)data;
+	if (mem && mem->mirrors) {
+		mem->mirrors->free = mem_free;
+		r_list_free (mem->mirrors);
+		mem->mirrors = NULL;
+	}
+	free (mem);
+}
+
+static void object_delete_items(RBinObject *o) {
+	ut32 i = 0;
+	r_return_if_fail (o);
+	sdb_free (o->addr2klassmethod);
+	r_list_free (o->entries);
+	r_list_free (o->fields);
+	r_list_free (o->imports);
+	r_list_free (o->libs);
+	r_list_free (o->relocs);
+	r_list_free (o->sections);
+	r_list_free (o->strings);
+	r_list_free (o->symbols);
+	r_list_free (o->classes);
+	r_list_free (o->lines);
+	sdb_free (o->kv);
+	if (o->mem) {
+		o->mem->free = mem_free;
+	}
+	r_list_free (o->mem);
+	o->mem = NULL;
+	o->entries = NULL;
+	o->fields = NULL;
+	o->imports = NULL;
+	o->libs = NULL;
+	o->relocs = NULL;
+	o->sections = NULL;
+	o->strings = NULL;
+	o->symbols = NULL;
+	o->classes = NULL;
+	o->lines = NULL;
+	o->info = NULL;
+	o->kv = NULL;
+	for (i = 0; i < R_BIN_SYM_LAST; i++) {
+		free (o->binsym[i]);
+		o->binsym[i] = NULL;
+	}
+}
+
 R_IPI void r_bin_object_free(void /*RBinObject*/ *o_) {
 	RBinObject *o = o_;
 	if (!o) {
 		return;
 	}
 	free (o->regstate);
 	r_bin_info_free (o->info);
-	r_bin_object_delete_items (o);
+	object_delete_items (o);
 	R_FREE (o);
 }
 
+static char *swiftField(const char *dn, const char *cn) {
+	char *p = strstr (dn, ".getter_");
+	if (!p) {
+		p = strstr (dn, ".setter_");
+		if (!p) {
+			p = strstr (dn, ".method_");
+		}
+	}
+	if (p) {
+		char *q = strstr (dn, cn);
+		if (q && q[strlen (cn)] == '.') {
+			q = strdup (q + strlen (cn) + 1);
+			char *r = strchr (q, '.');
+			if (r) {
+				*r = 0;
+			}
+			return q;
+		}
+	}
+	return NULL;
+}
+
+static RList *classes_from_symbols(RBinFile *bf, RBinObject *o) {
+	RBinSymbol *sym;
+	RListIter *iter;
+	RList *symbols = o->symbols;
+	RList *classes = o->classes;
+	if (!classes) {
+		classes = r_list_newf ((RListFree)r_bin_class_free);
+	}
+	r_list_foreach (symbols, iter, sym) {
+		if (sym->name[0] != '_') {
+			continue;
+		}
+		const char *cn = sym->classname;
+		if (cn) {
+			RBinClass *c = r_bin_class_new (bf, sym->classname, NULL, 0);
+			if (!c) {
+				continue;
+			}
+			// swift specific
+			char *dn = sym->dname;
+			char *fn = swiftField (dn, cn);
+			if (fn) {
+				// eprintf ("FIELD %s  %s\n", cn, fn);
+				RBinField *f = r_bin_field_new (sym->paddr, sym->vaddr, sym->size, fn, NULL, NULL);
+				r_list_append (c->fields, f);
+				free (fn);
+			} else {
+				char *mn = strstr (dn, "..");
+				if (mn) {
+					// eprintf ("META %s  %s\n", sym->classname, mn);
+				} else {
+					char *mn = strstr (dn, cn);
+					if (mn && mn[strlen (cn)] == '.') {
+						mn += strlen (cn) + 1;
+						// eprintf ("METHOD %s  %s\n", sym->classname, mn);
+						r_list_append (c->methods, sym);
+					}
+				}
+			}
+		}
+	}
+	if (r_list_empty (classes)) {
+		r_list_free (classes);
+		return NULL;
+	}
+	return classes;
```

Comment:
Maybe we need to fill o->classes directly here instead of returning a list, because this function is appending them and returning which is conceptually wrong and may lead to dupped data

Reply:
actually... it seems like classes will be always NULL :S I don't see any r_list_ on the `classes` variable in `classes_from_symbols`. Do you see where this is filled? It seems all this thing with classes is useless :/ I'd just keep the r_list_foreach in the body and remove the rest...

Reply:
ok